src/VecSim/algorithms/hnsw/hnsw_single_batch_iterator.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "hnsw_batch_iterator.h"
+
+template <typename DataType, typename DistType>
+class HNSWSingle_BatchIterator : public HNSW_BatchIterator<DataType, DistType> {
+private:
+    inline void fillFromExtras(candidatesLabelsMaxHeap<DistType> *top_candidates) override;
+    inline VecSimQueryResult_List prepareResults(candidatesLabelsMaxHeap<DistType> *top_candidates,
+                                                 size_t n_res) override;
+
+    inline void updateHeaps(candidatesLabelsMaxHeap<DistType> *top_candidates, DistType dist,
+                            idType id) override;
+
+public:
+    HNSWSingle_BatchIterator(void *query_vector, HNSWIndex<DataType, DistType> *index,
+                             VecSimQueryParams *queryParams,
+                             std::shared_ptr<VecSimAllocator> allocator)
+        : HNSW_BatchIterator<DataType, DistType>(query_vector, index, queryParams, allocator) {}
+
+    ~HNSWSingle_BatchIterator() override {}
+};
+
+/******************** Implementation **************/
+
+template <typename DataType, typename DistType>
+VecSimQueryResult_List HNSWSingle_BatchIterator<DataType, DistType>::prepareResults(
+    candidatesLabelsMaxHeap<DistType> *top_candidates, size_t n_res) {
+    VecSimQueryResult_List rl = {0};
+    // size_t initial_results_num = array_len(batch_results);
+    // Put the "spare" results (if exist) in the results heap.
+    while (top_candidates->size() > n_res) {
+        this->top_candidates_extras.emplace(top_candidates->top().first,
+                                            top_candidates->top().second); // (distance, label)
+        top_candidates->pop();
+    }
+    rl.results = array_new_len<VecSimQueryResult>(top_candidates->size(), top_candidates->size());
+    // Return results from the top candidates heap, put them in reverse order in the batch results
+    // array.
+    for (int i = (int)(top_candidates->size() - 1); i >= 0; i--) {
+        VecSimQueryResult_SetId(rl.results[i], top_candidates->top().second);
+        VecSimQueryResult_SetScore(rl.results[i], top_candidates->top().first);
+        top_candidates->pop();
+    }
+    return rl;
+}
+
+template <typename DataType, typename DistType>
+void HNSWSingle_BatchIterator<DataType, DistType>::fillFromExtras(
+    candidatesLabelsMaxHeap<DistType> *top_candidates) {
+    while (top_candidates->size() < this->ef && !this->top_candidates_extras.empty()) {
+        top_candidates->emplace(this->top_candidates_extras.top().first,
+                                this->top_candidates_extras.top().second);
+        this->top_candidates_extras.pop();
+    }
+}
+
+template <typename DataType, typename DistType>
+void HNSWSingle_BatchIterator<DataType, DistType>::updateHeaps(
```

Comment:
consider changing te ifs order
(first check cand.size < ef, else check the lower_bound > dist)